src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2265,107 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (ShouldRestoreNullValues())
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties = resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType);
+
+                        // All dynamic properties: null value restoration for properties that are not present.
+                        foreach (string name in resourceState.SelectedProperties.GetSelectedDynamicProperties(edmStructuredType))
+                        {
+                            if (resource.Properties.Any(p => p.Name.Equals(name, StringComparison.Ordinal))
+                                || resourceState.NavigationPropertiesRead.Contains(name))

```

Comment:
Dynamic property is represented by string; while regular declared property is represented by IEdmProperty.
For declared property, we need to check the IEdmProperty.Type.IsNullable; while for dynamic property, we don't.
Combining declared properties and dynamic properties together would require using strings to represent both; and then for each name we would need to try resolving it back to IEdmProperty to check Type.IsNullable; additionally, as we discussed, internal processing for entire-subtree and empty-subtree are for declared properties only, we would have to perform an additional step to build the list of selected dynamic properties to be combined with selected declared properties. With all these, it would not be efficient if we combining both kinds.

Therefore I think we should keep processing of above two kinds separate, and it is cleaner when we separate different processing logic.

---
In reply to: [229536394](https://github.com/OData/odata.net/pull/1106#discussion_r229536394) [](ancestors = 229536394)